src/v/iceberg/update_partition_spec_action.cc

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "iceberg/update_partition_spec_action.h"
+
+namespace iceberg {
+
+namespace {
+
+// If the key matches, the new field will be assigned the same id.
+struct field_key_t {
+    nested_field::id_t source_id;
+    transform transform;
+    ss::sstring name;
+
+    bool operator==(const field_key_t& other) const = default;
+
+    template<typename H>
+    friend H AbslHashValue(H h, const field_key_t& k) {
+        return H::combine(std::move(h), k.source_id, k.name);
+    }
+};
+
+} // namespace
+
+ss::future<action::action_outcome>
+update_partition_spec_action::build_updates() && {
+    updates_and_reqs ret;
+
+    const auto* cur_schema = table_.get_schema(table_.current_schema_id);
+    if (!cur_schema) {
+        co_return errc::unexpected_state;
+    }
+
+    const auto cur_spec_id = table_.default_spec_id;
+    auto highest_spec_id = partition_spec::id_t{-1};
+    chunked_hash_map<field_key_t, partition_field::id_t> field_key2id;
+    for (const auto& pspec : table_.partition_specs) {
+        highest_spec_id = std::max(highest_spec_id, pspec.spec_id);
+        for (const auto& field : pspec.fields) {
+            field_key2id[{field.source_id, field.transform, field.name}]
+              = field.field_id;
+        }
+    }
+
+    partition_spec resolved;
+    chunked_hash_set<ss::sstring> field_names_set;
+    const auto last_field_id = table_.last_partition_id;
+    auto cur_field_id = last_field_id + 1;
+    for (const auto& field : new_spec_.fields) {
+        const auto* source_field = cur_schema->schema_struct.find_field_by_name(
+          field.source_name);
+        if (!source_field) {
+            // failed to resolve
+            co_return errc::unexpected_state;
```

Comment:
nit: should we add some logging around the error handling to know what was wrong ?